include/inexor/vulkan-renderer/wrapper/device.hpp

```diff
@@ -0,0 +1,106 @@
+#pragma once
+
+#include "inexor/vulkan-renderer/availability_checks.hpp"
+#include "inexor/vulkan-renderer/settings_decision_maker.hpp"
+
+#include <spdlog/spdlog.h>
+#include <vulkan/vulkan.h>
+
+#include <cassert>
+#include <memory>
+#include <optional>
+#include <string>
+#include <vector>
+
+namespace inexor::vulkan_renderer::wrapper {
+
+/// @brief A RAII wrapper for VkDevice, VkPhysicalDevice and VkQueues.
+class Device {
+private:
+    VkDevice device = VK_NULL_HANDLE;
+    VkPhysicalDevice graphics_card = VK_NULL_HANDLE;
+    VkQueue graphics_queue = VK_NULL_HANDLE;
+    VkQueue present_queue = VK_NULL_HANDLE;
+    VkQueue transfer_queue = VK_NULL_HANDLE;
+    VkSurfaceKHR surface;
+
+    std::optional<std::uint32_t> present_queue_family_index = std::nullopt;
+    std::optional<std::uint32_t> graphics_queue_family_index = std::nullopt;
+    std::optional<std::uint32_t> transfer_queue_family_index = std::nullopt;
+
+    // TODO: Make proper use of queue priorities in the future.
+    const float global_queue_priority = 1.0f;
+
+    std::vector<VkDeviceQueueCreateInfo> queues_to_create = {};
+
+    AvailabilityChecksManager availability_checks;
+    VulkanSettingsDecisionMaker settings_decision_maker;
+
+    bool use_distinct_data_transfer_queue = false;
+    bool use_one_queue_for_graphics_and_present = false;
+
+    void prepare_queues(bool prefer_distinct_transfer_queue);
+
+public:
+    /// Delete the copy constructor so shaders are move-only objects.
+    Device(const Device &) = delete;
+    Device(Device &&other) noexcept;
+
+    /// Delete the copy assignment operator so shaders are move-only objects.
+    Device &operator=(const Device &) = delete;
+    Device &operator=(Device &&) noexcept = default;
+
+    /// @brief Creates a graphics card interface.
+    /// @param preferred_gpu_index [in] The index of the preferred physical device to use.
+    Device(const VkInstance &instance, const VkSurfaceKHR &surface, bool enable_vulkan_debug_markers, bool prefer_distinct_transer_queue,
```

Comment:
`instance` and `surface` don't need to be passed by reference.

Reply:
Right.

Reply:
Done.